### ISSUE TITLE
Database not on the correct thread ...

### ISSUE THREAD

**Issue by mtissington** [groue/GRDB.swift#55]
Sorry this is a little complicated ... I need to do a select from a couple of tables in one database and insert them into another. I'd like to do this in a transaction. At the moment my code is not doing that.
I have nested dbQueue (different) and this is causing a problem ...
Any sugestions please?

dbQueueMerge = source databaase
dbQueue = destination database

```swift
let dbQueueMerge = try! DatabaseQueue(path: url.path!, configuration: config)
        dbQueue.inDatabase { db in
            try! dbQueueMerge.inDatabase {dbMerge in
                for groupRow in Row.fetch(dbMerge, "SELECT * FROM groups") {
                    var newid: Int64 = 0
                    let groupid: Int64 = groupRow.value(named: "groupid")
                    let name: String = groupRow.value(named: "name")
                    let description: String = groupRow.value(named: "description")
                    try db.execute("INSERT INTO groups (name, description) VALUES (?, ?)",                       arguments: [name, description])
                    newid = db.lastInsertedRowID
                    for wordRow in Row.fetch(dbMerge, "SELECT * FROM words WHERE groupid = \(groupid)") {
                        let word: String = wordRow.value(named: "word")
                        let action: String = wordRow.value(named: "action")
                        try db.execute("INSERT INTO words (groupid, word, action) VALUES (?, ?, ?)",
                            arguments: [newid, word, action])
                    }
                }
            }
        }

**Comment by groue** [comment]
Would you please reformat your code as below:

<pre>
```swift
code...
```
</pre>

It will help a lot reading your question.

**Comment by mtissington** [comment]
yes, having a great difficult with code formatting ... seems to want to strip out tabs and cr/lf

**Comment by mtissington** [comment]
ok, that should do ..

**Comment by groue** [comment]
Thanks ;-)

OK now it's crystal clear. The inDatabase method executes its closure argument in a protected dispatch queue. And a database complains if it is used outside of this queue:

```swift
dbQueue.inDatabase { db in
    // In the queue of `db`
    try! dbQueueMerge.inDatabase {dbMerge in
        // In the queue of `dbMerge`
        // And there `db` complains it is not in its queue:
        try db.execute(...)
    }
}
```

OK. That's a pattern that is not supported today. I have to think about it.

Meanwhile, you have two possible workarounds:

1- First fetch all data in the source database queue, and then write in the destination database queue. It's much less memory efficient than your tight and sharp loop, I agree. But if you don't have too much data to copy, this solution can be an acceptable workaround.

2- If you actually copy the whole content of a database in another, use the [backup method](https://github.com/groue/GRDB.swift#backup).

**Comment by mtissington** [comment]
Hmm, I was afraid you would tell me something like that. I would have thought that merging a second database is a (fairly) common need. In the merge process PK values change so I'll have to go with option one.

Hope you can figure something out ... and thanks for the speedy replies!

**Comment by groue** [comment]
> I would have thought that merging a second database is a (fairly) common need.

There are my needs, the ones of my coworkers, and the ones of users like you who are kind enough to ask for a feature. Until now, only the need for backup has been expressed :-)

> Hope you can figure something out

So do I. Until this time, use a workaround, and stay tuned on GRDB :-)

**Comment by groue** [comment]
@mtissington 

This one is tricky to solve well.

First, why do we have those fatal errors? Because it's important to prevent the user to use a database connection out of its dispatch queue. There are many opportunities to make a concurrency mistake, and some of them are easy to trigger. GRDB actively prevents them with fatal errors:

```swift
let rows = dbQueue.inDatabase { db in
    Row.fetch(db, "SELECT * FROM table")
}
// Fatal error: a sequence can not be iterated outside of the database queue.
// Iterate an array returned by fetchAll() instead.
for row in rows { ... }
```

```swift
dbQueue.inDatabase { db in
    ...
    dispatch_async(dispatch_get_main_queue()) {
        // Fatal error: database wasn't used on the correct database queue
        let item = Item.fetchOne(db, key: 1)
    }
}
```

```swift
let cachedInsertStatement = try dbQueue.inDatabase { db in
    try db.updateStatement("INSERT INTO table (col) VALUES (?)")
}
// Fatal error: database wasn't used on the correct database queue
try cachedInsertStatement.execute(arguments: ["foo"])
```

Now, why is it tricky to provide a solution to your respectable request?

It's because the standard library [GCD](https://developer.apple.com/library/ios/documentation/Performance/Reference/GCD_libdispatch_Ref/) actively prevents introspection of the "queue stack". I don't know of any solid way for some code to know if it is being run inside a queue:

```swift
dispatch_sync(queue1) {
    dispatch_sync(queue2) {
        // Easy to know we're serialized in queue2.
        // But we don't know we're serialized in queue1.
    }
}
```

We could hack a solution on database queues:

```swift
dbQueue1.inDatabase {
    dbQueue2.inDatabase {
        // Easy to know we're serialized in dbQueue2.
        // Kind of possible to know we're serialized in dbQueue1.
    }
}
```

But introduce a raw dispatch_sync in the middle, and we're doomed:

```swift
dbQueue1.inDatabase {
    dispatch_sync(bombQueue) {
        dbQueue2.inDatabase {
            // We don't know we're serialized in dbQueue1.
        }
    }
}
```

That's what I mean by "solid". It's possible to provide a brittle solution, but I don't know how to provide a solid one, assuming it would be possible, which I doubt.

Maybe we could ask for an access to both databases at the same time:

```swift
DatabaseQueue.inDatabases(dbQueue, dbQueueMerge) { (db, dbMerge) in
    // both databases can be used here.
}
```

I don't quite know yet how to answer your request.

**Comment by groue** [comment]
Note for myself: learn about GCD "target queues"

**Comment by groue** [comment]
@mtissington OK I eventually managed to find a solution:

```swift
dbQueue1.inDatabase { db1 in 
    dbQueue2.inDatabase { db2 in 
        // Both db1 and db2 can be used here.
    }
}
```

**Comment by mtissington** [comment]
cool, will give it a try ....

**Comment by groue** [comment]
The fix has shipped in v0.66.0.

**Comment by mtissington** [comment]
confirmed that this now works - thanks

```swift
            try dbQueue.inTransaction { db in
                try dbQueueMerge.inDatabase {dbMerge in

```
